pkg/keyspace/tso_keyspace_group.go

```diff
@@ -72,42 +122,137 @@ func (m *GroupManager) GetKeyspaceGroupByID(id uint32) (*endpoint.KeyspaceGroup,
 		err error
 	)
 
-	m.store.RunInTxn(m.ctx, func(txn kv.Txn) error {
+	if err := m.store.RunInTxn(m.ctx, func(txn kv.Txn) error {
 		kg, err = m.store.LoadKeyspaceGroup(txn, id)
 		if err != nil {
 			return err
 		}
 		return nil
-	})
+	}); err != nil {
+		return nil, err
+	}
 	return kg, nil
 }
 
 // DeleteKeyspaceGroupByID deletes the keyspace group by id.
 func (m *GroupManager) DeleteKeyspaceGroupByID(id uint32) error {
-	m.store.RunInTxn(m.ctx, func(txn kv.Txn) error {
+	var (
+		kg  *endpoint.KeyspaceGroup
+		err error
+	)
+
+	if err := m.store.RunInTxn(m.ctx, func(txn kv.Txn) error {
+		kg, err = m.store.LoadKeyspaceGroup(txn, id)
+		if err != nil {
+			return err
+		}
 		return m.store.DeleteKeyspaceGroup(txn, id)
-	})
+	}); err != nil {
+		return err
+	}
+
+	userKind := endpoint.StringUserKind(kg.UserKind)
+	m.Lock()
+	defer m.Unlock()
+	// TODO: move out the keyspace to another group
+	// we don't need the keyspace group as the return value
+	m.groups[userKind].Remove(id)
+
 	return nil
 }
 
-func (m *GroupManager) saveKeyspaceGroups(keyspaceGroups []*endpoint.KeyspaceGroup) error {
+func (m *GroupManager) saveKeyspaceGroups(keyspaceGroups []*endpoint.KeyspaceGroup, overwrite bool) error {
 	return m.store.RunInTxn(m.ctx, func(txn kv.Txn) error {
 		for _, keyspaceGroup := range keyspaceGroups {
 			// TODO: add replica count
 			newKG := &endpoint.KeyspaceGroup{
-				ID:       keyspaceGroup.ID,
-				UserKind: keyspaceGroup.UserKind,
+				ID:        keyspaceGroup.ID,
+				UserKind:  keyspaceGroup.UserKind,
+				Keyspaces: keyspaceGroup.Keyspaces,
 			}
 			// Check if keyspace group has already existed.
 			oldKG, err := m.store.LoadKeyspaceGroup(txn, keyspaceGroup.ID)
```

Comment:
In UpdateKeyspaceGroup, we call m.saveKeyspaceGroups([]*endpoint.KeyspaceGroup{oldKG, newKG}, true). In this case, is it possible to skip m.store.LoadKeyspaceGroup and just call m.store.SaveKeyspaceGroup?

Reply:
Do you mean when overwrite is true, we directly SaveKeyspaceGroup?